### DISCUSSION TITLE
Update() performance

### DISCUSSION THREAD

**Discussion by FREEZX** [FirstGearGames/FishNet#746] [Feature Request]
Just adding on to an existing issue I noticed:
https://github.com/FirstGearGames/FishNet/issues/745
We are at a point where we're trying to squeeze out every bit of performance that may be on the table. I tried to change FishNet's updates to use a Manager instead of relying on Update() alone, and managed to shave off a small but cool ~0.3ms in my project (77 to 80 fps).
In our case I used an excellent but discontinued library called SwissArmyLib that we use for a bunch of stuff in the game - so I converted all FishNet classes that use Update to rely on their ManagedUpdateBehaviour - NetworkObject, NetworkBehaviour and NetworkTransform mainly.
https://swissarmylib-docs.readthedocs.io/en/latest/Events/ManagedUpdateBehaviour.html
The downside to this approach is that using this custom loop doesn't use the execution priority of the script file, but has support for a custom code-based priority. It also allows you to still use the normal Update() function of unity that respects execution priority, so final users wouldn't be affected by a change like this, but will get a little bit of performance from the Update change on the NetworkTransforms and NetworkObjects

**Comment by SnaiperoG3D** [comment]
Much more simplier fix for ppl who do not have libs. We had chat in DM in Discord with FGG. If someone needs it:
1. Fix for NA and NT. 
a) change Update method name to whatever you like.
b) subscribe to OnUpdate in OnStartClient, and unsubscribe in OnStopClient like so:
```c#
        public override void OnStartClient()
        {
            base.TimeManager.OnUpdate += CustomUpdate;
        }

        public override void OnStopClient()
        {
            if (base.TimeManager != null)
                base.TimeManager.OnUpdate -= CustomUpdate;
        }
```
2. Fix for NOB.
a) NetworkObject.cs - rename private void Update() to private void TimeManager_Update()
b) NetworkObject.Prediction.cs - goto method private void InvokeStartCallbacks_Prediction(bool asServer)
Change to this
```c#
        private void InvokeStartCallbacks_Prediction(bool asServer)
        {
            if (_predictionBehaviours.Count == 0)
                return;

            if (!asServer)
            {
                TimeManager.OnUpdate += TimeManager_Update;
                PredictionSmoother?.OnStartClient();
            }
        }
```
c) NetworkObject.Prediction.cs - goto method private void InvokeStopCallbacks_Prediction(bool asServer)   //right below last one.
change to this
```c#
        private void InvokeStopCallbacks_Prediction(bool asServer)
        {
            if (_predictionBehaviours.Count == 0)
                return;

            if (!asServer)
            {
                TimeManager.OnUpdate -= TimeManager_Update;
                PredictionSmoother?.OnStopClient();
            }
        }
```

In my case i have 5k nobs in the scene and in my working machine with Ryzen 1700 i have huge boost in performance, 6.41ms vs 2.31ms
Hard to catch exactly value cuz update node in profiler now too small

PS. The boost will be as big as the number of nobs you have in the scene

**Comment by FirstGearGames** [comment]
These changes will go live in 4.4.2.
![image](https://github.com/user-attachments/assets/5137d3e0-2937-4514-b2d3-fbb9d3bf72ef)
